### DISCUSSION TITLE
Proxmox VE Processor Microcode script update.

### DISCUSSION THREAD

**Discussion by tteck** [tteck/Proxmox#1794] [Announcements]
The Proxmox VE Processor Microcode script has been updated to allow users to select, download, and install available microcode packages.

Many have asked about the manual update of CPU microcode packages versus relying on the apt package manager and the non-free-firmware repository. 

I have a preference for manually managing certain updates, like the CPU microcode. Typically, after a new release becomes available, I take the time to thoroughly review the changelog. I also exercise a bit of caution and wait for a day or sometimes two, just to ensure there haven't been any reported issues from other users. If everything seems stable and reliable, I go ahead and apply the update.

However, I'm considering offering users the option to utilize the non-free-firmware repository for their apt updates, which could provide a more seamless, but less informed update experience. Thoughts?

**Comment by jcastro** [comment]
I've tried the script, installed the second option, restarted and it's showing the same date as before in the title of the alert (2021-11-12). Is this ok or should it be updated to the same date as the package? Thanks!

<img width="681" alt="image" src="https://github.com/tteck/Proxmox/assets/190036/c43c2928-90c0-4766-bc05-c8b310994266">

**Reply by tteck** [reply]
Not all CPU microcode patches are applicable to all CPU models. If your CPU wasn't part of the patch, then the microcode version won't change because it's not intended for your specific CPU.

Here's mine
![Screenshot 2023-09-14 1 59 20 PM](https://github.com/tteck/Proxmox/assets/21342670/bc130d3c-4902-40ad-877b-589decc76ba3)
As you can see, February.

**Reply by tteck** [reply]
[microcode-20230808 Release Notes](https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/releases/tag/microcode-20230808)

**Reply by jcastro** [reply]
ahh got it, I didn't know how it worked. Thanks!

**Comment by tteck** [comment]
It's important to note that downgrading microcode can have security implications, as newer microcode versions often include fixes for security vulnerabilities. Therefore, it's essential to assess the trade-offs between performance and security when choosing to install an older microcode package. If security is a concern, it's generally recommended to keep your microcode up to date whenever possible.

**Comment by rloomans** [comment]
Personally I prefer set and forget…. So count me in on auto updates.

**Comment by dan1el127** [comment]
I ran the script, selected the second option.
My date went from not available to 2023-02-23 
So i know it worked. Thanks!

**Comment by gabriel-gfs** [comment]
Could you add the -i flag to the grep as on some systems "microcode: microcode" can't find a result because the entry is "microcode: Microcode". I couldnt find where to submit for the script website ❤️

**Reply by tteck** [reply]
Using the `-i` option will yield an undesirable result.

Current code
```
journalctl -k | grep -E "microcode: microcode" | awk -F 'microcode: microcode updated early to revision |, date = ' '{print $2 ", date = " $3}'
0xf4, date = 2023-02-23
```
With the `-i` flag
```
journalctl -k | grep -Ei "microcode: microcode" | awk -F 'microcode: microcode updated early to revision |, date = ' '{print $2 ", date = " $3}'
0xf4, date = 2023-02-23
, date = 
```
Which systems make use of "microcode: Microcode"?

**Reply by gabriel-gfs** [reply]
I am using a Ryzen 3600 witha b350 motherboard and I get "microcode: Microcode"

**Reply by tteck** [reply]
Please verify if this snippet displays correctly.

```
journalctl -k | grep -iE "microcode: microcode" | awk -F 'microcode: microcode updated early to revision |, date = ' '{if (NF > 1) print $2 ", date = " $3}'
```

**Reply by gabriel-gfs** [reply]
`Zenbleed: please update your microcode for the most optimal fix` and `kernel: microcode: Microcode Update Driver: v2.2` as output of `journalctl -k | grep -iE "microcode"` so i guess this is an issue with my microcode installation although installing `amd64-microcode` reports to be on the newest version already. this is way beyond my understanding 😨

**Reply by tteck** [reply]
If you copy the code snippet above, and paste it in the Proxmox shell and hit enter, it should show you the microcode version and date. You asked for a fix, now I need to know if this "fix" resolves the issue on your Ryzen machine so that I can update the script.

**Reply by gabriel-gfs** [reply]
Sorry. No this doesnt result in any output

**Reply by tteck** [reply]
If I understand correctly, neither the existing code nor the "fix" I suggested generates any output. Is that accurate?

**Reply by tteck** [reply]
```
journalctl -k | grep -iE "microcode: microcode" | awk -F 'microcode: microcode updated early to revision |, date = ' '{if (NF > 1) print $2 ", date = " $3}'
```
So, this command essentially does the following:

- It searches for lines in the kernel messages that contain "microcode: microcode" (case-insensitive).
- For the lines that match, it extracts and prints the microcode revision and date, while filtering out lines where the date is missing or has unexpected content.

This command should provide you with the desired output format while handling case-insensitivity and filtering out unwanted lines.

Edit: The result obtained by executing the code in the Proxmox shell on my system is `0xf4, date = 2023-02-23`
